### PR TITLE
make test count iteration configurable by environment variable

### PR DESCRIPTION
**Pull request by apoelstra**
You can set the test iteration count by command-line parameter but it is very hard to control the iteration count used by `make check` and even harder to control the count when this is run through Bitcoin Core's build system. Using an environment var solves both problems.

### CODE REVIEW

**Comment by real-or-random** [comment]
Nice, that's indeed useful. 

ACK 0ce45548813709d828cb3abcc7db4c9ce6e26907 I inspected the diff

Here are a few small annoying things that you can add to this PR if you want. If not, let's consider this a list for another PR:
 * Add check that aborts if not `count > 0`? Technically `count == 0` is fine but I think noone wants to run this on purpose as it just disables some tests arbitrary. 
 * I was tempted to say that we should have a `ITERS` also for the exhaustive tests but I think that's not true. The default of 2 makes a lot of sense for these, because the randomness is only used to randomize the representation in jacobian coordinates. But we should add the same `count > 0` check there.
 * Document the environment variables. Since we don't have a .md for tests, the canonical place is in the "usage" output. (This is not optimal because the benchmarks have an env variable but no "usage" output. Yeah, ok.)

**Review by jonasnick — APPROVED** [review]
ACK 0ce45548813709d828cb3abcc7db4c9ce6e26907

**Comment by elichai** [comment]
utACK 0ce45548813709d828cb3abcc7db4c9ce6e26907

> * Add check that aborts if not `count > 0`? Technically `count == 0` is fine but I think noone wants to run this on purpose as it just disables some tests arbitrary.

I also wanted to do some sanitizing on the bench env var but hadn't gotten to doing it yet
https://github.com/bitcoin-core/secp256k1/blob/ca4906b02e69644d83e13b3c09cc9147fc1232a5/src/bench.h#L124

**Comment by apoelstra** [comment]
deadalnix was using count==0 :) https://github.com/bitcoin-core/secp256k1/issues/528

**Comment by elichai** [comment]
> deadalnix was using count==0 :) #528
lol
I didn't fix yet because I think those env vars are for "advanced users only" and only tests/benchmarks so it doesn't matter too much

**Comment by apoelstra** [comment]
I'm happy to add a count>0 check if people think it's valuable.

I don't think it's worth exposing a way to change the exhaustive test iteration count. I didn't even know there was one.

Unfortunately there is no usage text for the tests binary :). I could add an unconditional message above the "count =" line.

**Comment by sipa** [comment]
The count variable for the exhaustive tests changes the z coordinate for gej elements (which is randomized, not exhaustive).

I don't think there is much need to changing the default, though.

**Comment by real-or-random** [comment]
> deadalnix was using count==0 :) #528

yeah but even he says:
> My concern here isn't really that 0 is a very useful thing to test for, but that it put the system is some weird state that reveal a bug either in libsecp256k1 or possibly in gcc's optimizer. 

So I think we should just enforce > 0. But either way is fine.

> I don't think there is much need to changing the default, though.

Indeed.

File: src/tests.c [hunk 1]
```diff
@@ -5626,6 +5626,15 @@ int main(int argc, char **argv) {
     /* find iteration count */
     if (argc > 1) {
         count = strtol(argv[1], NULL, 0);
+    } else {
+        const char* env = getenv("SECP256K1_TEST_ITERS");
+        if (env) {
+            count = strtol(env, NULL, 0);
+        }
+    }
+    if (count == 0) {
```
Comment (real-or-random) [inline src/tests.c:5635]: Maybe just
```suggestion
    if (count <= 0) {
```

**Review by real-or-random — APPROVED** [review]
ACK a729a6c6bfb07929c56222fc43cba790d9e5c083 I inspected the diff but didn't test

File: src/tests.c [hunk 2]
```diff
@@ -5626,6 +5626,15 @@ int main(int argc, char **argv) {
...
     if (argc > 1) {
         count = strtol(argv[1], NULL, 0);
+    } else {
+        const char* env = getenv("SECP256K1_TEST_ITERS");
+        if (env) {
+            count = strtol(env, NULL, 0);
+        }
+    }
+    if (count <= 0) {
+        fputs("An iteration count of 0 is not allowed.\n", stderr);
```
Comment (jonasnick) [inline src/tests.c:5636]: nit: `<= 0`

**Review by jonasnick — COMMENTED** [review]
ACK mod nit

**Review by jonasnick — APPROVED** [review]
ACK f4fa8d226a95e42b252c07edb425c446370e01c0

**Review by real-or-random — APPROVED** [review]
ACK f4fa8d226a95e42b252c07edb425c446370e01c0

**Pull request merged**